File: LeetCode/Medium/3201-Find_The_Maximum_Length_Of_Valid_Subsequence_I/3201-FindTheMaximumLengthOfValidSubsequenceI.ex
defmodule Solution do
  @spec maximum_length(nums :: [integer]) :: integer
  def maximum_length(nums) do
    initial_state = %{
      odd: 0,
      even: 0,
      alt_even: 0,
      alt_odd: 0,
      next_even: true,
      next_odd: true
    }

    final_state = Enum.reduce(nums, initial_state, fn num, acc ->
      is_odd = rem(num, 2) != 0

      new_acc = Map.merge(acc, %{
        odd: if(is_odd, do: acc.odd + 1, else: acc.odd),
        even: if(not is_odd, do: acc.even + 1, else: acc.even)
      })

      new_acc =
        if is_odd do # Current number is ODD
          if acc.next_odd do
            Map.put(new_acc, :alt_odd, acc.alt_odd + 1)
            |> Map.put(:next_odd, false)
          else
            new_acc
          end
        else # Current number is EVEN
          if not acc.next_odd do
            Map.put(new_acc, :alt_odd, acc.alt_odd + 1)
            |> Map.put(:next_odd, true)
          else
            new_acc
          end
        end

      new_acc =
        if not is_odd do # Current number is EVEN
          if acc.next_even do
            Map.put(new_acc, :alt_even, acc.alt_even + 1)
            |> Map.put(:next_even, false)
          else
            new_acc
          end
        else # Current number is ODD
          if not acc.next_even do
            Map.put(new_acc, :alt_even, acc.alt_even + 1)
            |> Map.put(:next_even, true)
          else
            new_acc
          end
        end

      new_acc
    end)

    Enum.max([
      final_state.odd,
      final_state.even,
      final_state.alt_odd,
      final_state.alt_even
    ])
  end
end

File: LeetCode/Medium/3201-Find_The_Maximum_Length_Of_Valid_Subsequence_I/3201-FindTheMaximumLengthOfValidSubsequenceI.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumLength = function (nums) {
    let odd = 0;
    let even = 0;
    let alternate_even = 0;
    let alternate_odd = 0;
    let next_even = true;
    let next_odd = true;

    for (let num of nums) {
        if (num & 1) {
            // Odd
            odd += 1;
            if (next_odd) {
                alternate_odd += 1;
                next_odd = false;
            }

            if (!next_even) {
                alternate_even += 1;
                next_even = true;
            }
        } else {
            // Even
            even += 1;

            if (next_even) {
                alternate_even += 1;
                next_even = false;
            }

            if (!next_odd) {
                alternate_odd += 1;
                next_odd = true;
            }
        }
    }

    return Math.max(...[even, odd, alternate_even, alternate_odd]);
};